File: dynamic_connectivity/WeightedQuickUnionUF.java

/*
 * Weighted quick-union
 * - Modify quick-union to avoid tall trees.
 * - Keep track of size of each tree (number of objects).
 * - Balance by linking root of smaller tree to root of bigger tree.
 */


public class WeightedQuickUnionUF
{
  private final int[] id;
  private final int[] treeSize;  
  private int numConnected;

  public WeightedQuickUnionUF(int numComponents)
  {
    id = new int[numComponents];
    treeSize = new int[numComponents];
    numConnected = numComponents;

    for (int i = 0; i < numComponents; i++)
    {
      id[i] = i;
      treeSize[i] = 1;
    }

  }


  public int find(int component)
  {
    while (id[component] != component)
    { 
      id[component] = id[id[component]]; // path compression
      component = id[component];
    }
    return component;
  }

  /*
   * p and q are connected iff they have the same root
   */
  public boolean connected(int p, int q)
  {
    return find(p) == find(q);
  }

  /*
   * change the root of p to that of q
   */
  public void union(int p, int q)
  {
    final int rp = find(p);
    final int rq = find(q);

    if (rp == rq) return;

    if (treeSize[rp] < treeSize[rq])
    {
      id[rp] = rq;
      treeSize[rq] += treeSize[rp];
    }
    else
    {
      id[rq] = rp;
      treeSize[rp] += treeSize[rq];
    }

    --numConnected;
  }

  private void print(final int[] array, final String label)
  {
    final int padding = label.length();
    for (int i=0; i<padding+2+(2*array.length); i++)
      System.out.printf("=");
    System.out.println("");

    System.out.printf("%-" + (padding+2) + "s", " ");
    for (int i = 0; i < array.length; i++)
      System.out.printf("%3d", i);
    System.out.println("");

    System.out.printf("%-" + padding + "s%s", label, "[]");
    for (int i = 0; i < array.length; i++)
      System.out.printf("%3d", array[i]);
    System.out.println("");
  }

  public void print()
  {
    print(id, "id");
    print(treeSize, "tree size");
  }

  public static void main(String[] args)
  {
    WeightedQuickUnionUF qf = new WeightedQuickUnionUF(10);
    qf.union(4,3);
    qf.union(3,8);
    qf.union(6,5);
    qf.union(9,4);
    qf.union(2,1);
    qf.union(5,0);
    qf.union(7,2);
    qf.union(6,1);
    qf.union(7,3);
    qf.print();
   
    /* ======= Expected Result =======
    index 0  1  2  3  4  5  6  7  8  9
     id[] 6  2  6  4  6  6  6  6  4  4
    */
  }
}

File: dynamic_connectivity/WeightedQuickUnionUF.rb
#!/usr/bin/env ruby

# Implementation of Weighted Quick-union (lazy approach) from the Algorithms course from
# Princeton University on Coursera
# Initialisation: N
# Find: log_2 N
# Union: log_2 N (includes cost of finding roots)

class WeightedQuickUnionUF
  attr_reader :id     # for rspec
  attr_reader :count  # for rspec
  attr_reader :tree_size # for rspec

  def initialize(num_elements)
    @id = Array.new(num_elements){|i|i}
    @tree_size = Array.new(num_elements){|i|1}
    @count = num_elements
  end

  # link root of smaller tree to root of larger tree
  # update the tree_size array of the root
  def union(p, q)
    root_p = find(p)
    root_q = find(q)

    unless root_p == root_q
      if @tree_size[root_p] < @tree_size[root_q]
        @id[root_p] = root_q
        @tree_size[root_q] += @tree_size[root_p]
      else
        @id[root_q] = root_p
        @tree_size[root_p] += @tree_size[root_q]
      end
      @count -= 1
    end
  end

  def connected(p, q)
    find(p) == find(q)
  end

  def find(i)
    while i != @id[i]
      @id[i] = @id[@id[i]] ## path compression
      i = @id[i]
    end
    return i
  end

  def print_id
    @id.each_with_index {|item, index| print " #{index} "}
    print "\n#{@id}\n"
  end

end

